### PR TITLE
Predict total consumption by fuel type

### PR DESCRIPTION
**Pull request by mikivee**
1. Predicting by HVAC consumption by fuel type rather than by end use. See [description](https://www.notion.so/rewiringamerica/Building-towards-an-MVP-9cb748ec5ef647fba50a302fa7f152ca?pvs=4#975bcd6c5cbe4dd7b170ca65fc48b901).
2. Predicting total consumption for all upgrades (excluding standalone range and dryer upgrades). See [description](https://www.notion.so/rewiringamerica/Building-towards-an-MVP-9cb748ec5ef647fba50a302fa7f152ca?pvs=4#70bea834972b410e9e2a17c5463e44a3)

### CODE REVIEW

File: src/databricks/model_evaluation.py
Comment (mikivee) [inline src/databricks/model_evaluation.py:1]: Ugh, sorry this is gonna make this more annoying to review, but it looks like the diff was large enough that with the rename, it is treating this as a whole add/delete file situation 😢 .

File: src/databricks/model_training.py
Comment (mikivee) [inline src/databricks/model_training.py:88]: moved to datagen.py

File: src/databricks/model_evaluation.py
Comment (mikivee) [inline src/databricks/model_evaluation.py:255]: similar to before

Comment (mikivee) [inline src/databricks/model_evaluation.py:255]: very similar to before

File: src/databricks/model_evaluation.py
Comment (mikivee) [inline src/databricks/model_evaluation.py:283]: very similar to before

File: src/databricks/model_evaluation.py
Comment (mikivee) [inline src/databricks/model_evaluation.py:358]: same as before

File: src/databricks/model_evaluation.py
Comment (mikivee) [inline src/databricks/model_evaluation.py:428]: pretty much the same as before, but adding total_metrics_by_fuel df.

**Review by tarrenpeterson — APPROVED** [review]
This looks great to me - I know there was a lot of work happening behind the scenes to roll back that final layer of the layer cake related to the custom loss function but you made it look easy. I have a couple small questions but the thorough documentation preempted most questions

File: src/databricks/model_training.py
Comment (mikivee) [reply src/databricks/model_training.py:235]: Good question. So before we had `layers.Dense(64,  activation = "leaky_relu")(bm)` which first applies the linear mapping and then the non-linear activation. But since we want to apply batchnorm in between the linear map and the activation, we have to instead not apply any non-linear activation at first (i.e., activation='linear' which is the default), then pass it thru a batchnorm layer and then pass it through the leakyrelu layer. There were two additional things that made this confusing that I have now fixed to make this more readable and easier to modify in the future:
1. when you have repeated "blocks" like this (e.g, conv-batchnorm-relu) the standard thing is to define a function for this block rather than just repeat it over and over. I was just being lazy. I have now done this. 
2. I was still using this pattern of manually applying the activation after the dense layer in the combine trunk, even tho I wasn't using batchnorm, so I could have just passed the activation name directly. I removed the activation from the arguments here since this is kind of stupid now, and then updated layers in combine trunk to use the simpler pattern.

File: src/databricks/model_training.py
Comment (mikivee) [reply src/databricks/model_training.py:311]: yeah that is fair... I think the reason I put it there is because failed at inference time for like 1000 different reasons so far, and most of the time, it means I did something wrong in training and have to retrain the model. So I figured it would be good if it failed there just in case I don't run the eval script and I think "oh yay successful model run" and then don't realize I need to retrain. I can nest it in the DEBUG if you feel strongly.

File: src/databricks/model_evaluation.py
Comment (mikivee) [reply src/databricks/model_evaluation.py:553]: My sense was that this was mostly just due to the fact that the savings are larger for 4 than 3, and so when using these relative measures, 4 looks better than 3. If you look at the error for just the consumption rather than savings  [here](https://docs.google.com/spreadsheets/d/1XO9a_NPOcqaZrsisJSCs5T3R2enVSPJ-eKI1Ilgf-H8/edit?gid=86053945#gid=86053945) (the right half), you'll see that they are quite similar. But there may be something deeper going on and I just don't realize it..

**Review by mikivee — COMMENTED** [review]
Thanks @tarrenpeterson! Super helpful and thorough comments as always. Gonna assign back to you to just look over the unresolved comments if you wanna take a look at those before I merge.

File: src/databricks/build_feature_store.py
Comment (tarrenpeterson) [reply src/databricks/build_feature_store.py:477]: No the docstring is sufficient - sorry for not catching that on first pass

File: src/databricks/model_training.py
Comment (tarrenpeterson) [reply src/databricks/model_training.py:235]: Ok that makes sense - I'd missed that detail about the ordering of steps here and that the batchnorm function needed to be added in between the linear mapping and non-linear activation. Thanks for adding the function in as well!

File: src/databricks/model_training.py
Comment (tarrenpeterson) [reply src/databricks/model_training.py:311]: I don't feel strongly on this point so will defer to your judgement here

File: src/databricks/model_evaluation.py
Comment (tarrenpeterson) [reply src/databricks/model_evaluation.py:106]: I think it's a great place to start without doing any work on optimization yet

File: src/databricks/model_evaluation.py
Comment (tarrenpeterson) [reply src/databricks/model_evaluation.py:553]: Ahh that's a good call - I didn't think of it from that point of view

File: src/databricks/model_training.py
Comment (tarrenpeterson) [reply src/databricks/model_training.py:101]: It would make sense to make it the new default to me but I don't think it's a big deal either way

File: src/databricks/datagen.py
Comment (tarrenpeterson) [reply src/databricks/datagen.py:465]: Ahh I understand the first point and did not consider that when making the suggestion that you subset for `n_train` or `n_test` before the train/test split. But agree about the second point here and thank you for correcting with the ceiling!

File: src/databricks/model_training.py
Comment (mikivee) [reply src/databricks/model_training.py:101]: ya u right. i'll update.

File: src/databricks/model_training.py
Comment (mikivee) [reply src/databricks/model_training.py:311]: okay I'm going to keep for now

**Pull request merged**
